### ISSUE TITLE
defaultValue callback is never called

### ISSUE THREAD

**Issue by gpietro** [marmelab/react-admin#4800]
**What you were expecting:**
I'm trying to setting a default uuid on a TextInput field as suggested in the documentation.

```
<TextInput
    disabled
    source='id'
    defaultValue={() => uuidv4()}
/>
```

but it seems that the callback is never called and the field is empty even on form submit.
By doing the following:

```
<TextInput
    disabled
    source='id'
    defaultValue={uuidv4()}
/>
```

The defaultValue appears in the TextInput but in an ArrayInput the value would be always the same for each Array element because uuid4() is called only once.

**Environment**

* React-admin version: ^3.5.0
* React version: ^16.13.1
* Browser: Chrome 81 - Mac

**Comment by fzaninotto** [comment]
Thanks for reporting this. The bug is in the **documentation**, and unfortunately what you ask is no longer possible. 

Since version 3, react-admin switched from `redux-form` to `react-final-form` to handle forms. Redux-form didn't handle field-level default values, so the job was done by react-admin - and we added support for function values. React-final-form supports field default values natively, so we removed the react-admin code for default values in v3. And unfortunately, react-final-form doesn't support functions for initial values. 

You'll have to use `React.useMemo` unstead. And we need to update the documentation to remove the part saying that functions are accepted as default values for inputs.

**Comment by gpietro** [comment]
I tried and unfortunately even with React.useMemo doesn't work for ArrayInput.

```
<Create title='Create form' {...props}>
		<SimpleForm>
			<TextInput
				disabled
				source='id'
				label='id'
				defaultValue={React.useMemo(() => uuidv4(), [])}
			/>
			<TextInput source='name' />
			<ArrayInput source='wards'>
				<SimpleFormIterator>
					<TextInput
						disabled
						source='id'
						label='id'
						defaultValue={React.useMemo(() => uuidv4(), [])}
					/>
					<TextInput source='name' label='Name' />					
				</SimpleFormIterator>
			</ArrayInput>
		</SimpleForm>
	</Create>
```

**Comment by fzaninotto** [comment]
The `defaultValue` for `ArrayInput` may have a problem, regardless of the functional value. Please build a reproducible test case using CodeSandbox and open a new issue if you can reproduce it.

**Comment by gpietro** [comment]
I'm having trouble using react-admin locally with npm link so I'll try here.
Could it be the missing 
```
defaultValue: input.props.defaultValue
```
while cloning element the origin of the problem?

https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/form/SimpleFormIterator.js#L153

**Comment by gpietro** [comment]
I'm having trouble using react-admin locally with npm link so I'll try here.
Could it be the missing 
```
defaultValue: input.props.defaultValue
```
while cloning element the origin of the problem?

https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/form/SimpleFormIterator.js#L153

**Comment by fzaninotto** [comment]
No, `cloneElement` clones an element with its props.
